### PR TITLE
Akm/register files

### PR DESCRIPTION
**Pull request by akm303**
creating CPU and basis of register related components

### CODE REVIEW

File: src/main/java/group1/mips_simulator/components/README.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+## Emulation Component of Simulator
+
```
Comment (toon-leader-bacon) [inline src/main/java/group1/mips_simulator/components/README.md:2]: Whats the plan with this README.md file? Will it be filled out later? Or should this be removed?

File: src/main/java/group1/mips_simulator/components/Register.java [hunk 2]
```diff
@@ -5,6 +5,13 @@
  * the CPU to read/ write quickly.
  */
 public class Register {
+    regType type;
+    public Value value;
+
+    private enum regType{
```
Comment (toon-leader-bacon) [inline src/main/java/group1/mips_simulator/components/Register.java:11]: Usually type names start with a capitol letter. Not critical, but just pointing it out for future reference.

File: src/main/java/group1/mips_simulator/components/README.md (hunk 1 above)
Comment (toon-leader-bacon) [reply src/main/java/group1/mips_simulator/components/README.md:2]: Oh I see, this is a hold over from the other branch. See my comment on that other MR then :)

File: src/main/java/group1/mips_simulator/components/Register.java [hunk 3]
```diff
@@ -5,6 +5,13 @@
...
+        GP, //General Purpose Registers
+        IX; //IndeX Registers
+    }
 
     public Value read() {
         // TODO
```
Comment (toon-leader-bacon) [inline src/main/java/group1/mips_simulator/components/Register.java:17]: What's the plan with these TODO statements? Are you planning on addressing these or is that a different task? It should be pretty easy right, just `return this.value;`? Or is there more complexity

Comment (toon-leader-bacon) [reply src/main/java/group1/mips_simulator/components/Register.java:17]: Same question wit the `set(...)` function

File: src/main/java/group1/mips_simulator/components/RegisterFile.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package group1.mips_simulator.components;
+import group1.mips_simulator.components.Register;
+
+public class RegisterFile {
```
Comment (toon-leader-bacon) [inline src/main/java/group1/mips_simulator/components/RegisterFile.java:4]: Can you add some documentation to this class? What is a "RegisterFile" class represent? Just a sentence or two describing what it is. 

How is this distinct from the `MipsComputer.registers` vector?

File: src/main/java/group1/mips_simulator/components/RomLoader.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package group1.mips_simulator.components;
+
+public class RomLoader {
+    /*
+    read the boot program from the ROM and place it into memory in a location you designated.
```
Comment (toon-leader-bacon) [inline src/main/java/group1/mips_simulator/components/RomLoader.java:5]: Any idea on where in memory we should put the program? I'll suggest exactly at the half way point in our memory. So like $0 -> $half is for storage and $(half + 1) -> $end  is for program instructions. But that's just a suggestion I'm open to other ideas.

File: src/main/java/group1/mips_simulator/components/CPU.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package group1.mips_simulator.components;
+
+public class CPU {
+    RegisterFile Registers = new RegisterFile();
```
Comment (toon-leader-bacon) [inline src/main/java/group1/mips_simulator/components/CPU.java:4]: Will the CPU be a component of the MipsComputer class? Like will the computer "have a" CPU? 

Also, variable names are usually starting with a lower case letter. So consider the following change:
```suggestion
    RegisterFile registers = new RegisterFile();
```

File: src/main/java/group1/mips_simulator/components/Register.java (hunk 2 above)
Comment (akm303) [reply src/main/java/group1/mips_simulator/components/Register.java:11]: i see, i can switch those;
So, classes, objects, types, all capitalized?

File: src/main/java/group1/mips_simulator/components/Register.java (hunk 3 above)
Comment (akm303) [reply src/main/java/group1/mips_simulator/components/Register.java:17]: yea, ill address those, just been putting together the skeleton, havent gotten down to implementing just yet.
was still processing the layout and project requirements, but ill be implementing these today

File: src/main/java/group1/mips_simulator/components/RegisterFile.java (hunk 4 above)
Comment (akm303) [reply src/main/java/group1/mips_simulator/components/RegisterFile.java:4]: can do

File: src/main/java/group1/mips_simulator/components/RomLoader.java (hunk 5 above)
Comment (akm303) [reply src/main/java/group1/mips_simulator/components/RomLoader.java:5]: He's suggested a location in the project description:
Under 3.1 Central Processor
...
NOTE: Thus, the first address of your program should be greater than your length of the program.
I suggest you load your boot program beginning at location octal 10 and the first address of your program at octal address = octal 10 + length of boot program + octal 10
...

I'm still deciphering what exactly this means. I dont know if its helpful, but im aware that the standard memory management architecture is usually something like:
[heap]. //dynamic variables
[.  v   ]
[.  -   ]
[.  ^  ]
[stack] //paramaters
[static] //global/static data
[code] //instructions

I dont know how true to this we need to stick for the project, so lemme think on that
tbc not anti the suggestion, just not sure what profs looking for

File: src/main/java/group1/mips_simulator/components/CPU.java (hunk 6 above)
Comment (akm303) [reply src/main/java/group1/mips_simulator/components/CPU.java:4]: woops, will fix the capitalizations

So, the way im thinking of it, and i have a suggestion for you on your end too (i can make it a more formal request after this, ill find the actual branch where you made this):

- could you rename the "MipsComputer" class to just "Computer"?

- a Computer utilizes a set of Hardware components (things like CPU, Memory, Busses, etc.) to operate on data
- each of these hardware components have their own sets of hardware, subcomponents, and operations they need to perform (i.e. the CPU contains registers, basic instruction set, a ROM loader; CPU needs to perform elements to execute the Instr set, read ROM contents to memeory, etc.) These are the attributes and responsibilities of the CPU
- I envision the Computer is almost just a collection of these components, and maybe the interface to the Operator+FieldEngi Consoles (or UI); 
i.e. we use the Computer class to configure our computer ("I want 1 cpu, 1 memory, 1 L2 cache) and be its interface to the UI, we'd define User operations and route them through here

this is just how im thinking of it, im very open to alternative architecture

File: src/main/java/group1/mips_simulator/components/Register.java (hunk 2 above)
Comment (toon-leader-bacon) [reply src/main/java/group1/mips_simulator/components/Register.java]: Yes! Although in java all those three things are the same. Classes are objects which create a new type. Different languages have different style conventions. Python for example

File: src/main/java/group1/mips_simulator/components/CPU.java (hunk 6 above)
Comment (toon-leader-bacon) [reply src/main/java/group1/mips_simulator/components/CPU.java]: Sounds good to me! Just wanna make sure I'm keeping up with the architecture as it evolves :)

**Review by toon-leader-bacon — APPROVED** [review]
Please address the merge conflicts, but otherwise looks good! Approved

**Pull request merged**